File: Sample/2.0/Controllers/Sample/OrderDetailsController.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNet.OData;
using Microsoft.AspNet.OData.Routing;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Internal;

namespace MyApp.Controllers.Sample
{
  using Models;
  using Data;
  using Models.Sample;

  [EnableQuery]
  [ODataRoutePrefix("odata/Sample/OrderDetails")]
  public partial class OrderDetailsController : Controller
  {
    private Data.SampleContext context;

    public OrderDetailsController(Data.SampleContext context)
    {
      this.context = context;
    }
    // GET /odata/Sample/OrderDetails
    [HttpGet]
    public IEnumerable<Models.Sample.OrderDetail> GetOrderDetails()
    {
      var items = this.context.OrderDetails.AsQueryable<Models.Sample.OrderDetail>();

      this.OnOrderDetailsRead(ref items);

      return items;
    }

    partial void OnOrderDetailsRead(ref IQueryable<Models.Sample.OrderDetail> items);

    [HttpGet("{Id}")]
    public IActionResult GetOrderDetail(int key)
    {
        var item = this.context.OrderDetails.Where(i=>i.Id == key).SingleOrDefault();

        if (item == null)
        {
            return NotFound();
        }

        return new ObjectResult(item);
    }
    partial void OnOrderDetailDeleted(Models.Sample.OrderDetail item);

    [HttpDelete("{Id}")]
    public IActionResult DeleteOrderDetail(int key)
    {
        var item = this.context.OrderDetails
            .Where(i => i.Id == key)
            .SingleOrDefault();

        if (item == null)
        {
            return NotFound();
        }

        this.OnOrderDetailDeleted(item);
        this.context.OrderDetails.Remove(item);
        this.context.SaveChanges();

        return new NoContentResult();
    }

    partial void OnOrderDetailUpdated(Models.Sample.OrderDetail item);

    [HttpPut("{Id}")]
    public IActionResult PutOrderDetail(int key, [FromBody]Models.Sample.OrderDetail newItem)
    {
        if (newItem == null || newItem.Id != key)
        {
            return BadRequest();
        }

        this.OnOrderDetailUpdated(newItem);
        this.context.OrderDetails.Update(newItem);
        this.context.SaveChanges();

        return new NoContentResult();
    }

    [HttpPatch("{Id}")]
    public IActionResult PatchOrderDetail(int key, [FromBody]JObject patch)
    {
        var item = this.context.OrderDetails.Where(i=>i.Id == key).FirstOrDefault();

        if (item == null)
        {
            return BadRequest();
        }

        Data.EntityPatch.Apply(item, patch);

        this.OnOrderDetailUpdated(item);
        this.context.OrderDetails.Update(item);
        this.context.SaveChanges();

        return new NoContentResult();
    }

    partial void OnOrderDetailCreated(Models.Sample.OrderDetail item);

    [HttpPost]
    public IActionResult Post([FromBody] Models.Sample.OrderDetail item)
    {
        if (item == null)
        {
            return BadRequest();
        }

        this.OnOrderDetailCreated(item);
        this.context.OrderDetails.Add(item);
        this.context.SaveChanges();

        return Created($"odata/Sample/OrderDetails/{item.Id}", item);
    }
  }
}

File: Sample/2.0_VB/Controllers/Sample/OrderDetailsController.vb
Imports Newtonsoft.Json.Linq
Imports Microsoft.AspNetCore.Mvc
Imports Microsoft.AspNet.OData
Imports Microsoft.AspNet.OData.Routing

Namespace MyApp.Controllers.Sample

    <EnableQuery>
    <ODataRoutePrefix("odata/Sample/OrderDetails")>
    Partial Public Class OrderDetailsController
        Inherits Controller
        Private context As Data.SampleContext

        Public Sub New(context As Data.SampleContext)
            Me.context = context
        End Sub

        <HttpGet>
        Public Function GetOrderDetails() As IEnumerable(Of Models.Sample.OrderDetail)
            Dim items = Me.context.OrderDetails.AsQueryable()

            Me.OnOrderDetailsRead(items)

            Return items
        End Function

        Partial Private Sub OnOrderDetailsRead(ByRef items As IQueryable(Of Models.Sample.OrderDetail))
        End Sub

        <HttpGet("{Id}")>
        Public Function GetOrderDetail(key As Integer) As IActionResult
            Dim item = Me.context.OrderDetails.Where(Function(i) i.Id = key).SingleOrDefault()

            If item Is Nothing Then
                Return NotFound()
            End If

            Return New ObjectResult(item)
        End Function

        Partial Private Sub OnOrderDetailDeleted(item As Models.Sample.OrderDetail)
        End Sub

        <HttpDelete("{Id}")>
        Public Function DeleteOrderDetail(key As Integer) As IActionResult
            Dim item = Me.context.OrderDetails.Where(Function(i) i.Id = key).SingleOrDefault()

            If item Is Nothing Then
                Return NotFound()
            End If

            Me.OnOrderDetailDeleted(item)
            Me.context.OrderDetails.Remove(item)
            Me.context.SaveChanges()

            Return New NoContentResult()
        End Function

        Partial Private Sub OnOrderDetailUpdated(item As Models.Sample.OrderDetail)
        End Sub

        <HttpPut("{Id}")>
        Public Function PutOrderDetail(key As Integer, <FromBody> newItem As Models.Sample.OrderDetail) As IActionResult
            If newItem Is Nothing OrElse newItem.Id <> key Then
                Return BadRequest()
            End If

            Me.OnOrderDetailUpdated(newItem)
            Me.context.OrderDetails.Update(newItem)
            Me.context.SaveChanges()

            Return New NoContentResult()
        End Function

        <HttpPatch("{Id}")>
        Public Function PatchOrderDetail(key As Integer, <FromBody> patch As JObject) As IActionResult
            Dim item = Me.context.OrderDetails.Where(Function(i) i.Id = key).FirstOrDefault()

            If item Is Nothing Then
                Return BadRequest()
            End If

            Data.EntityPatch.Apply(item, patch)

            Me.OnOrderDetailUpdated(item)
            Me.context.OrderDetails.Update(item)
            Me.context.SaveChanges()

            Return New NoContentResult()
        End Function

        Partial Private Sub OnOrderDetailCreated(item As Models.Sample.OrderDetail)
        End Sub

        <HttpPost>
        Public Function Post(<FromBody> item As Models.Sample.OrderDetail) As IActionResult
            If item Is Nothing Then
                Return BadRequest()
            End If

            Me.OnOrderDetailCreated(item)
            Me.context.OrderDetails.Add(item)
            Me.context.SaveChanges()

            Return Created("odata/Sample/OrderDetails/{item.Id}", item)
        End Function

    End Class

End Namespace
